src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs

```diff
@@ -964,16 +976,22 @@ private sealed class ReturnInferenceCacheKey
         {
             public readonly ImmutableArray<TypeWithAnnotations> ParameterTypes;
             public readonly ImmutableArray<RefKind> ParameterRefKinds;
+            public readonly ImmutableArray<DeclarationScope> ParameterDeclarationScopes;
             public readonly NamedTypeSymbol? TaskLikeReturnTypeOpt;
 
-            public static readonly ReturnInferenceCacheKey Empty = new ReturnInferenceCacheKey(ImmutableArray<TypeWithAnnotations>.Empty, ImmutableArray<RefKind>.Empty, null);
+            public static readonly ReturnInferenceCacheKey Empty = new ReturnInferenceCacheKey(ImmutableArray<TypeWithAnnotations>.Empty, ImmutableArray<RefKind>.Empty,
+                parameterDeclaredScopes: default, taskLikeReturnTypeOpt: null);
 
-            private ReturnInferenceCacheKey(ImmutableArray<TypeWithAnnotations> parameterTypes, ImmutableArray<RefKind> parameterRefKinds, NamedTypeSymbol? taskLikeReturnTypeOpt)
+            private ReturnInferenceCacheKey(ImmutableArray<TypeWithAnnotations> parameterTypes, ImmutableArray<RefKind> parameterRefKinds,
+                ImmutableArray<DeclarationScope> parameterDeclaredScopes, NamedTypeSymbol? taskLikeReturnTypeOpt)
             {
                 Debug.Assert(parameterTypes.Length == parameterRefKinds.Length);
+                Debug.Assert(parameterDeclaredScopes.IsDefault || parameterTypes.Length == parameterDeclaredScopes.Length);
```

Comment:
>parameterDeclaredScopes.IsDefault [](http://example.com/codeflow?start=29&length=33)

What is the scenario when this condition is true?

Reply:
I'll add today.

Reply:
When all parameters are unscoped, we use a `default` array.

Reply:
Created corresponding spec update: https://github.com/dotnet/csharplang/pull/6551